runtime/src/crowdfund.rs

```diff
@@ -0,0 +1,539 @@
+	deploy_data: Option<(Hash, Vec<u8>)>,
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as Example {
+		/// Info on all of the funds.
+		Funds get(funds):
+			map FundIndex => Option<FundInfo<T::AccountId, BalanceOf<T>, T::Hash, T::BlockNumber, ParaIdOf<T>>>;
+
+		/// The total number of funds that have so far been allocated.
+		FundCount get(fund_count): FundIndex;
+
+		/// The funds that have had additional contributions during the last block. This is used
+		/// in order to determine which funds should submit updated bids.
+		NewRaise: Vec<FundIndex>;
+
+		/// True if the fund was ending at the last block.
+		WasEnding: bool;
+	}
+}
+
+decl_event!(
+	pub enum Event<T>
+		where
+	<T as system::Trait>::AccountId,
+	Balance = BalanceOf<T>,
+	ParaId = ParaIdOf<T>,
+	{
+		Created(FundIndex),
+		Contributed(AccountId, FundIndex, Balance),
+		Withdrew(AccountId, FundIndex, Balance),
+		Retiring(FundIndex),
+		Dissolved(FundIndex),
+		DeployDataFixed(FundIndex),
+		Onboarded(FundIndex, ParaId),
+	}
+);
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event<T>() = default;
+		
+		/// Create a new crowdfunding campaign for a parachain slot deposit.
+		#[weight = TransactionWeight::Basic(100_000, 10)]
+		fn create(
+			origin,
+			#[compact] end: T::BlockNumber,
+			#[compact] cap: BalanceOf<T>,
+			#[compact] first_slot: T::BlockNumber,
+			#[compact] last_slot: T::BlockNumber
+		) {
+			let owner = ensure_signed(origin)?;
+			let deposit = T::SubmissionDeposit::get();
+
+			let imb = T::Currency::withdraw(
+				&owner,
+				deposit,
+				WithdrawReason::Transfer,
+				ExistenceRequirement::AllowDeath,
+			)?;
+
+			let index = FundCount::mutate(|c| { let r = *c; *c += 1; r });
+
+			// No fees are paid here if we need to create this account; that's why we don't just
+			// use the stock `transfer`.
+			T::Currency::resolve_creating(&Self::fund_account_id(index), imb);
+
+			<Funds<T>>::insert(index, FundInfo {
+				parachain: None,
+				owner: owner,
+				deposit: deposit,
+				raised: Zero::zero(),
+				end: end,
+				cap: cap,
+				// Ensure it's Some, so that the first contribution causes it to be inserted into
+				// `NewRaise`.
+				last_contribution: Some(Zero::zero()),
+				first_slot: first_slot,
+				last_slot: last_slot,
+				deploy_data: None,
+			});
+
+			Self::deposit_event(RawEvent::Created(index));
+
+		}
+		
+
+		
+		/// Contribute to a crowd sale. This will transfer some balance over to fund a parachain
+		/// slot. It will be withdrawable in two instances: the parachain becomes retired; or the
+		/// slot is
+		fn contribute(origin, #[compact] index: FundIndex, #[compact] value: BalanceOf<T>)
+		{
+			let who = ensure_signed(origin)?;
+
+			ensure!(value >= T::MinContribution::get(), "contribution too small");
+
+			let mut fund = Self::funds(index).ok_or("invalid fund index")?;
+			fund.raised += value;
+			ensure!(fund.raised <= fund.cap, "contributions exceed cap");
```

Comment:
It seems a fund can be trapped between two requirements:
* Contribution must make the total raised to be less than or equal to the fund cap (this)
* Contribution must be more than the min contribution

It seems like we should do something like:
* Before contributing, make sure that the total fund raised at the time is less than the fund cap
* Make sure value contributing is more than min contribution
* Only add the amount of funds needed to make it reach the fund cap, even if it is less than min contribution

In this case, there may be one contributor with an amount smaller than the min contribution, but only one, which means that the rule below is only broken one time. Probably not a big deal at that point.

```
//! Contributions must be of at least `MinContribution` (to account for the resources taken in
//! tracking contributions)
```

Reply:
who care if it can't quite 100% optimise its cap by a piece of dust?!

this is a complete non-issue. leave it alone.